### PR TITLE
Refactoring of ZCL_ABAPGIT_ENVIRONMENT

### PR DESCRIPTION
**Pull request by ThomasPloski**
Refactoring is needed in order to be able to mock this central abapGIT environment class in unit tests, so that different system states can be tested within unit tests.

### CODE REVIEW

**Comment by larshp** [comment]
constructor should be public, see https://github.com/SAP/styleguides/blob/master/clean-abap/CleanABAP.md#if-your-global-class-is-create-private-leave-the-constructor-public

**Comment by ThomasPloski** [comment]
Hi Lars,
Based on our discussion I shifted the instanciation into the ZCL_ABAPGIT_FACTORY class. Furthermore I use the ZCL_ABAPGIT_INJECTOR class as possibility to mock data during unit tests.

File: src/zcl_abapgit_environment.clas.locals_imp.abap [hunk 1]
```diff
@@ -0,0 +1,67 @@
+*"* use this source file for the definition and implementation of
+*"* local helper classes, interface definitions and type
+*"* declarations
+
+CLASS ltcl_abapgit_environment_stub DEFINITION FOR TESTING.
```
Comment (larshp) [inline src/zcl_abapgit_environment.clas.locals_imp.abap]: this is also a FOR TESTING class, suggest moving it to the testclass include

**Comment by larshp** [comment]
also new method `is_restart_required` is added, which is not used from anywhere, guess this is SCP relevant. It will show up in static analysis as unused, so suggest adding a comment to make sure its not deleted

**Comment by larshp** [comment]
for `ltcl_test_simple_positive` and `ltcl_test_simple_negative` it looks like none of the code in the class under test is tested? So it tests the factory and the injector, plus the stub implementation? The responsibility for testing the factory and injector can be moved to the factory and injector. And then the stub can be removed?

**Comment by ThomasPloski** [comment]
> for `ltcl_test_simple_positive` and `ltcl_test_simple_negative` it looks like none of the code in the class under test is tested? So it tests the factory and the injector, plus the stub implementation? The responsibility for testing the factory and injector can be moved to the factory and injector. And then the stub can be removed?

Testing the actual cut is quite hard because this class is just a wrapper to call different APIs to get some system states. Unfortunately,  there are no interfaces in place which whould allowed to mock this real API calls. For that reason I have created this stub including unit tests for reuse purpose in many different places in other unit tests.

**Comment by ThomasPloski** [comment]
> also new method `is_restart_required` is added, which is not used from anywhere, guess this is SCP relevant. It will show up in static analysis as unused, so suggest adding a comment to make sure its not deleted

In the next step I would contribute the usage of those methods in the abapGit core coding. But this will be done in some different pull requests.

**Comment by larshp** [comment]
the stub is a local class so it cannot be reused

I dont remember exactly the definition of what a "stub" is, but I guess it should not implement the class under test? But instead simulate behavior of something that the class under test depends on

**Comment by larshp** [comment]
source branch is deleted, @ThomasPloski will open a new PR with the changes

closing

**Pull request closed without merging**